v4l2loopback.c

```diff
@@ -1717,7 +1734,18 @@ static int vidioc_qbuf(struct file *file, void *fh, struct v4l2_buffer *buf)
 			b->buffer.timestamp = buf->timestamp;
 			b->buffer.flags |= V4L2_BUF_FLAG_TIMESTAMP_COPY;
 		}
-		b->buffer.bytesused = buf->bytesused;
+		if (dev->pix_format_has_valid_sizeimage) {
+			if (buf->bytesused >= dev->pix_format.sizeimage) {
+				b->buffer.bytesused = dev->pix_format.sizeimage;
```

Comment:
Side note, the pix_format sizeimage is meant to be the padded size, so that you can use a third party allocator like DMABuf Heap. It seems that the sizeimage here is the "ideal" image size without padding. Maybe it has nothing to do, but this could also be the reason of the initial issue.

Reply:
> Side note, the pix_format sizeimage is meant to be the padded size, so that you can use a third party allocator like DMABuf Heap. It seems that the sizeimage here is the "ideal" image size without padding. Maybe it has nothing to do, but this could also be the reason of the initial issue.

I don't know where this notion comes from but AFAIK this simply is *not* true (and never has been for the 14 years I've worked on v4l2). If you need a size with which is a multiple of the page-size then you need to do the rounding up to the next page-size yourself.

https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/pixfmt-v4l2.html describes the `sizeimage` field as:

```
Size in bytes of the buffer to hold a complete image, set by the driver. Usually this is bytesperline times height. When the image consists of variable length compressed data this is the number of bytes required by the codec to support the worst-case compression scenario.

The driver will set the value for uncompressed images.

Clients are allowed to set the sizeimage field for variable length compressed data flagged with V4L2_FMT_FLAG_COMPRESSED at [ioctl VIDIOC_ENUM_FMT](https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-enum-fmt.html#vidioc-enum-fmt), but the driver may ignore it and set the value itself, or it may modify the provided value based on alignment requirements or minimum/maximum size requirements. If the client wants to leave this to the driver, then it should set sizeimage to 0.
```

No where does this say that this is padded / round up to the next page-size. The rounding up to page-size is done for the length field of mmap-ed buffers since mmaping is done per page.

Reply:
It has been documented that way since bytesperline is what most driver will align. I believe I already said page alignment is uncommon strict requirement. I take note to improve the doc for driver implementer, but to make external allocation work (if that matter to you) software like chromium need that to be accurate. Its quite likely though that page alignment is always obtained by accident. As for what size is that actual data needed, every application dealing with raw video should know this, there is no need for the driver to tell us.